django/db/models/sql/compiler.py

```diff
@@ -2012,19 +2028,23 @@ def execute_sql(self, result_type):
         non-empty query that is executed. Row counts for any subsequent,
         related queries are not available.
         """
-        cursor = super().execute_sql(result_type)
-        try:
-            rows = cursor.rowcount if cursor else 0
-            is_empty = cursor is None
-        finally:
-            if cursor:
-                cursor.close()
+        if result_type not in {ROW_COUNT, NO_RESULTS}:
+            raise ValueError(f"Unknown cursor type for update {repr(result_type)}")
```

Comment:
This check serves as documentation (notice how `result_type == SINGLE` in previous versions of this code would blow up at `cursor.rowcount`)

Reply:
If this is being added, and you've added a release note documenting it, it should be tested

Reply:
I removed the assertion here (didn't feel strongly about that one, and there is a real cost).

I have this vague dissatisfaction about the end result of this, the combo of subclassing `execute_sql` in ways that totally give different signatures, the outdated docstring on `Compiler.execute_sql`, the result type polymorphism based on `result_type`... I figured out what was going on by staring at stuff but I wonder if there is a more straightforward version of this.

Put another way, I don't think that the `SQLCompiler` layout would pass a 2024 Django code review. But I don't have a very satisfying answer and now that I've looked at this I have what I needed in the exercise.